### PR TITLE
Ping

### PR DESCRIPTION
**Pull request by NiteBlock**

### CODE REVIEW

File: src/commands/mod.rs [hunk 1]
```diff
@@ -24,3 +29,103 @@ pub trait Command {
...
+    }
+}
+
+pub struct Cooldowns(Arc<Mutex<Vec<Cooldown>>>);
+
+impl Cooldowns {
+    pub fn new() -> Cooldowns {
+        Cooldowns(Arc::new(Mutex::new(Vec::new())))
+    }
+    async fn remove_cooldown(inner: Arc<Mutex<Vec<Cooldown>>>, cooldown: Cooldown) {
```
Comment (dtomvan) [inline src/commands/mod.rs:63]: Why not use `&self` instead of `inner`?

File: src/commands/mod.rs [hunk 2]
```diff
@@ -24,3 +29,103 @@ pub trait Command {
...
+        sleep(cooldown.duration).await;
+        let mut c = inner.lock().await;
+        let mut x = (&*c).clone();
+        for (i, v) in x.iter().enumerate() {
+            if v == &cooldown {
+                x.remove(i);
+                break;
+            }
+        }
+        *c = x
```
Comment (dtomvan) [inline src/commands/mod.rs:73]:
```suggestion
        if let Some(i) = c.iter().position(|v| v == &cooldown) {
            c.swap_remove(i);
        }
```

File: src/commands/mod.rs [hunk 3]
```diff
@@ -24,3 +29,103 @@ pub trait Command {
...
+                CooldownType::Global => true,
+                CooldownType::User(uid) => uid == user,
+            };
+            if x {
+                return Some(
+                    cooldown.duration - (cooldown.exipre - cooldown.duration).elapsed().unwrap(),
+                );
+            }
+        }
+        None
```
Comment (dtomvan) [inline src/commands/mod.rs:122]:
```suggestion
        self.0.lock().await.iter().find(|cooldown| {
            if key != cooldown.key && cooldown.key.is_some() {
                return false;
            };
            match cooldown.cooldown_info {
                CooldownType::Global => true,
                CooldownType::User(uid) => uid == user,
            }
        })
        .map(|cooldown| {
            cooldown.duration - (cooldown.exipre - cooldown.duration).elapsed().unwrap()
        })
```

**Review by dtomvan — CHANGES_REQUESTED** [review]
I've done some nitpicking

File: src/commands/mod.rs [hunk 4]
```diff
@@ -24,3 +29,103 @@ pub trait Command {
     where
         Self: Sized;
 }
+#[derive(Clone, Copy, PartialEq)]
```
[the remark this answers is not in the document]
Comment (NiteBlock) [reply src/commands/mod.rs:32]: ok

File: src/commands/mod.rs [hunk 5]
```diff
@@ -24,3 +29,103 @@ pub trait Command {
...
+        Cooldown {
+            duration,
+            exipre: SystemTime::now() + duration,
+            key,
+            cooldown_info,
+        }
+    }
+}
+
+pub struct Cooldowns(Arc<Mutex<Vec<Cooldown>>>);
```
[the remark this answers is not in the document]
Comment (NiteBlock) [reply src/commands/mod.rs:57]: I think so

File: src/commands/mod.rs (hunk 1 above)
Comment (NiteBlock) [reply src/commands/mod.rs:63]: it drops the value

File: src/commands/mod.rs [hunk 6]
```diff
@@ -24,3 +29,103 @@ pub trait Command {
...
+#[derive(Clone, Copy, PartialEq)]
+pub enum CooldownType {
+    User(UserId),
+    Global,
+}
+
+#[derive(Clone, PartialEq)]
+pub struct Cooldown {
+    duration: Duration,
+    exipre: SystemTime,
```
[the remark this answers is not in the document]
Comment (NiteBlock) [reply src/commands/mod.rs:41]: oops

**Review by dtomvan — APPROVED** [review]

**Comment by NiteBlock** [comment]
Oh thx I was gonna do it today

**Pull request merged**